### PR TITLE
Add elliptic utility to work with tags containing maps

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

These are utility functions to work with tags containing maps.
They will be used in the elliptic code to retrieve values from
maps in the DataBox, e.g. for boundary conditions which depend
on background fields that are sliced to the boundaries and kept
in persistent memory.

I use them to apply boundary conditions in #2732.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

**Comment by nilsdeppe** [comment]
This is inconsistent with the way boundary conditions are done and will be done in the evolution. Specifically, in the evolution there isn't really anything stored in maps anymore other than the unnormalized normal vector, and that has to change for curved meshes too. This is intentionally to avoid storing everything that's projected to boundaries in the DataBox (a very heavy penalty on memory usage). If you have fewer than 4-5 points per dimension, then storing the quantities on the boundaries/interfaces in the DataBox basically doubles your memory usage, in addition to not being cache-friendly.

If you insist on the design you have here for the elliptic code I'd prefer this functionality be added only to the elliptic code until we see a use for it in the evolution code.

**Comment by nilsvu** [comment]
I thought about this a lot but couldn't come up with great alternatives. Here's my "wish list":

- The boundary conditions should have access to background quantities sliced to element faces

  Many geometric quantities like face normals, coords and Jacobians as well as many background fields like Christoffel symbols, extrinsic curvature etc. don't change between steps in an elliptic solve, so they should be sliced to faces during initialization and stored in the DataBox as `DirectionMap`s. (In the case of a Gauss-Lobatto grid the slices could be views into the volume data without having to store extra data.) I think such background quantities also exist in evolutions, though perhaps fewer.
- The derived boundary conditions classes should list `argument_tags` (all required background quantities) and `volume_tags` (the subset of background quantities taken from the element volume, not the face).

  This means the boundary conditions implementations don't have to worry about how the slices are stored in the DataBox, i.e. in a `domain::Tags::Interface` or in a Variables with views or some other way, but they just list what they need on the element face. The code that calls the boundary conditions, typically an action, is responsible for retrieving things out of the DataBox.

  Making the interface for derived boundary conditions classes as easy as possible is important because it's a common way to add different physics to a simulation.

  As an alternative to listing `argument_tags` and `volume_tags` we could list `volume_tags` and `face_tags`, but it's probably better to be consistent with other parts of the code where we use `argument_tags` and `volume_tags`. In those other parts of the code this separation allows a class be applicable to volume data _and_ face data by just ignoring the `volume_tags` when applied to the volume (not needed for boundary conditions of course, just good to keep the interfaces consistent). Splitting the tag lists differently doesn't alleviate the need for the unmapping utility anyway.
- The `unmap_arg` helper struct that decides whether to unmap or pass through an argument, as well as the detail function that expands the `tmpl::list`s to parameter packs (i.e. the code that's currently in `ApplyAt.hpp` in this PR), shouldn't be duplicated in every file that calls boundary conditions (currently that's 3 files: the standard DG operator action, the action that imposes inhomogeneous boundary conditions on the RHS and the subdomain operator. Also other files deal with maps in the DataBox).

Possible alternatives to `db::apply_at`:

- Require boundary conditions classes to list the full DataBox tag like `domain::Tags::Interface<domain::Tags::BoundaryConditionsExterior<Dim>, domain::Tags::Coordinates<Dim, Frame::Inertial>>` instead of just `domain::Tags::Coordinates<Dim, Frame::Inertial>`.

  This puts more responsibility on the person writing the boundary conditions class and makes it harder to change how the data is stored in the DataBox. It also means the `direction` has to be passed in to the boundary conditions as a special argument just so the class can retrieve things `at` the direction. A full map is passed in to the boundary conditions even though the class only needs the data for a particular key.
- Move `db::apply_at` to `Domain/InterfaceHelpers.hpp`

   The unmapping utility works with any (nested) map in the DataBox, so it's not limited to interface tags. In fact, we want more flexibility than storing everything in interface tags, so I thought it would be good to make this utility not specific to interface tags.
- Move to `Elliptic/?`

  This problem is not specific to elliptic solves, also evolutions need to deal with sliced background quantities. If you want to deal with that differently, I'd be happy to hear how, so I might adopt the same approach. Even if we decide to do different things in elliptic vs evolution to see which works out best, I would use this utility not only in boundary conditions, so I can't think of a good place to put it (top-level `Elliptic/ApplyAt.hpp` seems wrong).
- Move to `Utilities/Tuples/ApplyAt.hpp`

  We could move some generic tuple utility that applies to any of our tagged containers, like `DataBox`, `TaggedTuple` and `Variables`, to `Utilities/Tuples/`. This would be like `std::apply` works not only on `std::tuple` but other similar types, like `std::array` and `std::pair`.

That's all I can think of right now. I'm happy to hear alternative suggestions. Just trying to get the interface right and write code that's easiest to maintain :)

**Comment by nilsdeppe** [comment]
Thanks for the details!

- If you only need these quantities for boundary conditions this won't be of
  much concern, but if you are storing the quantities in the inside of the domain on
  internal faces then you should be aware that with 6 grid points per dimension
  you double your memory usage, fewer it more than doubles, and at 12 points per
  dimension it adds a 50% overhead. This is a lot of persistent memory, but of
  course the tradeoff is fewer CPU operations/slicing.

  In the evolution case there are almost none of these quantities. Boundary
  conditions are generally time-dependent, the only background fields would be
  the metric with GR+matter evolutions. However, since this scales with the
  boundaries of the domain, I expect that in most or all interesting simulations
  the cost of recomputing the background spacetime on the boundary will be very
  cheap compared to the actual hydro evolution.

  I don't know how you can have views with GL points. The volume data is stored
  as xi-varies-fastest, so you can only have views on boundaries where the
  normal is in the zeta (I think) direction.
- I don't know how much of the non-interface compute tags of the code use
  `volume_tags`. I guess boundary corrections do, but I can't think of anything
  else right now. I'm not sure on why exactly listing `volume_tags` is superior
  to listing `face_tags`, or explicitly listing what you expect from a face or a
  volume. Having said that, I don't think any of these approaches is
  unreasonable, but do keep in mind that the evolution code will stop using
  interface tags completely very soon and at that point I think the interface
  tags should be move to the elliptic directory.
- I'm not sure to what extent I agree with this. If the elliptic code uses this
  sort of approach, then I think the code should be in the elliptic
  directory. `apply_at` shouldn't exist generally and the reason we have
  interface tags is because they don't actually change the DataBox functionality
  (db::apply_at does because it's being added as a db:: utility). It is not the
  DataBox's job to change or manipulate types. We currently do this for
  unique_ptr but whether that's a good decision or not is something @kidder and
  I keep going back and forth on.

In terms of the alternatives:
- Reasonable, but consider not using interface tags at all if you can. Don't you
  need these quantities on the interior side anyway? In the evolution code I
  just copy such quantities from the interior side, making the boundary
  correction list what is allowed to be copied from the interior.
- Reasonable as long as the function is not in the `db` namespace (I'm not sure
  what a reasonable namespace actually is, though).
- Reasonable, or could do `Elliptic/Utilities/ApplyAt.hpp`
  (e.g. elliptic::apply_at)
- I don't see how this is a good idea. std::apply doesn't do anything more than
  unpack the tuple. I don't see a good reason to start adding a lot of niche
  helper functions for specific classes. I can't see it scaling well and will
  very likely lead to a lot of difficult to understand code.

I would prefer the 3rd alternative (or 1st if it's possible). This is not a
utility that should be used outside the elliptic code.

**Comment by nilsvu** [comment]
Notes from a call with @nilsdeppe today:

- Elliptic solves are more static than evolutions (duh 🤯). The evolution code might end up storing no slices in persistent memory at all, while the elliptic code does, it's not entirely clear yet what will turn out the most efficient tradeoff between CPU and memory cost. @nilsdeppe prefers the unmapping utility be moved to `Elliptic/` and only move to more general scope if needed.

**Comment by nilsvu** [comment]
@nilsdeppe This is now moved to `Elliptic/` as you suggested. Please take another look. @kidder Could you do the second review?

File: src/Elliptic/Utilities/ApplyAt.hpp [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+      const gsl::not_null<T*> arg,
+      const std::tuple<MapKeys...>& /*map_keys*/) noexcept {
+    return arg;
+  }
+};
+
+template <typename FirstMapKey, typename... MapKeys>
+struct unmap_arg<false, FirstMapKey, MapKeys...> {
+  template <typename T>
+  static constexpr auto apply(
```
Comment (nilsdeppe) [inline src/Elliptic/Utilities/ApplyAt.hpp]: are you sure this (and below) sholudn't be a `decltype(auto)`?

File: tests/Unit/Elliptic/Utilities/Test_ApplyAt.cpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Framework/TestingFramework.hpp"
+
+#include <map>
+#include <string>
+#include <tuple>
+#include <unordered_map>
```
Comment (nilsdeppe) [inline tests/Unit/Elliptic/Utilities/Test_ApplyAt.cpp:9]: Do you want test this with a `DirectionMap` as well? I don't see why it wouldn't work, but might not be bad to test. Up to you!

**Review by nilsdeppe — COMMENTED** [review]
Feel free to squash any changes

File: src/Elliptic/Utilities/ApplyAt.hpp (hunk 1 above)
Comment (nilsvu) [reply src/Elliptic/Utilities/ApplyAt.hpp]: added `decltype`s. `DataBox` might be missing some as well

File: tests/Unit/Elliptic/Utilities/Test_ApplyAt.cpp (hunk 2 above)
Comment (nilsvu) [reply tests/Unit/Elliptic/Utilities/Test_ApplyAt.cpp:9]: added

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**